cmd/snap-update-ns/user.go

```diff
@@ -33,24 +36,51 @@ type UserProfileUpdateContext struct {
 	// the update operation is occurring. It may be the current UID but doesn't
 	// need to be.
 	uid int
+	// home contains the user's real home directory
+	home string
+}
+
+// isPlausibleHome returns and error if the path is not clean and absolute path to a directory
+// (symlinks not followed) that can be opened for reading by the user.
+func isPlausibleHome(path string) error {
+	if path != filepath.Clean(path) {
+		return fmt.Errorf("cannot allow unclean path")
+	}
+	if !filepath.IsAbs(path) {
+		return fmt.Errorf("cannot allow relative path")
+	}
+	const openFlags = syscall.O_NOFOLLOW | syscall.O_CLOEXEC | syscall.O_DIRECTORY
+	fd, err := sysOpen(path, openFlags, 0)
+	if err != nil {
+		return err
+	}
+	sysClose(fd)
+	return nil
 }
 
 // NewUserProfileUpdateContext returns encapsulated information for performing a per-user mount namespace update.
-func NewUserProfileUpdateContext(instanceName string, fromSnapConfine bool, uid int) *UserProfileUpdateContext {
+func NewUserProfileUpdateContext(instanceName string, fromSnapConfine bool, uid int) (*UserProfileUpdateContext, error) {
+	realHome := os.Getenv("SNAP_REAL_HOME")
+	if realHome == "" {
+		return nil, fmt.Errorf("cannot retrieve home directory")
+	}
+	// See bootstrap.c function switch_to_privileged_user(). When snap-update-ns is invoked for
+	// user mounts, the effective uid and gid is changed to the calling user and supplementary
+	// groups dropped, while retaining capability SYS_ADMIN. Having the effective uid and gid
+	// changed to the calling user is a prerequisite for isPlausibleHome to function as intended.
+	if err := isPlausibleHome(realHome); err != nil {
+		return nil, fmt.Errorf("cannot use invalid home directory %q: %v", realHome, err)
```

Comment:
I am reconsidering failing early, because it is possible for users to exist without home directories specified, which means snap-update-ns might start failing, even for someone who is not using personal-files or this feature.

If we do not fail early, the problem is that when real home is not available because it does not exist or is invalid, then we end up with mount profile entries that will attempt to create dirs '/$HOME'/... which in all probability will just fail (which solves the problem) because /'$HOME' does not expected to exist, however, it seems more appropriate to avoid attempts to perform actions on unexpanded mount entries at all, which fundamentally leaves us with 2 approaches:

1. filter out/remove all unexpanded mount entries from the profile early when the profile is loaded, and indicate this was done with WARNING: ...
2. modify mount entry to indicate that it should not be acted on by any `changePerformImpl` step because of expansion issue, this could be new option `x-snapd.invalid=$HOME unavailable or invalid`. What I prefer about this approach, it would give a consistent experience flagging problematic entries [here at perform time](https://github.com/snapcore/snapd/blob/master/cmd/snap-update-ns/update.go#L85). We can populate the error using the message in option `x-snapd.invalid=...`

@pedronis, your view on this would be appreciated.

Reply:
Great catch, I more inclined towards the `x-snapd.invalid` approach because it be used for other entries later, but in both approaches we need to consider how to act when users create their homedirs afterwards.

- How NeededChanges will handle entries with x-snapd.invalid in case the homedirs exist?
- Will the user have to wait for an operation (e.g. refresh, connection) that triggers mount entries recalculations (NeededChanges), or can we have something detect this on each run?

Reply:
I'm not entirely a fan of creating entries that are just marked invalid,  later phases need to sort things etc, but need to ignore them etc etc. 

> How NeededChanges will handle entries with x-snapd.invalid in case the homedirs exist?

if we don't have them but they now appear, this is just more work to do that it should be able to handle

Reply:
Elaborating more on the idea in case it may be useful:

A desired mount entry marked as "invalid"/"unexpanded", should not be treated different from any other desired entry while calculating neededChanges. It serves as an instruction that performing the change (in this case ensuring the target) should not be attempted and must fail.

This means it will be treated the same as any other failed needed change, it will NEVER become part of the list of changes that is saved and loaded as "current" changes at the next invocation.

Reply:
It turns out that, regardless of the new snap-update-ns early fail, if the calling user home directory does not exist, snap-confine will error (after returning from snap-update-ns) and the snap application will not execute:

(1) snap cmd run warns during snap run stage:
`cmd_run.go:1046: WARNING: cannot create user data directory: cannot create snap home dir: mkdir /home/ernest3: permission denied`
(2) Error at snap-confine stage, because it cannot create the user directory.
`cannot create user data directory: /home/ernest3/snap/test-personal-files/x1: Permission denied`
See: 
- https://github.com/snapcore/snapd/blob/master/cmd/snap-confine/snap-confine.c#L461-L465
- [dependency on $SNAP_USER_DATA](https://github.com/snapcore/snapd/blob/master/cmd/snap-confine/user-support.c#L29) (same prefix than $SNAP_REAL_HOME)

So in conclusion, the new snap-update-ns early fail, will only change where it fails, but not that it fails.

=> We could consider adding a check for the existence of $SNAP_REAL_HOME earlier in snap-confine, before running snap-update-ns for user mount.

Reply:
@perdonis, here is my interpretation of the concept you suggested:
(1) Still [check validity of user home](https://github.com/snapcore/snapd/pull/13342/files#diff-5862fc64285517730e1ae794ad4f4c455c57533d287e3da20d3110ec6b4fa5d2R71-R79) when creating user update context
(2) [Store the home (valid/invalid/empty) and also the error](https://github.com/snapcore/snapd/pull/13342/files#diff-5862fc64285517730e1ae794ad4f4c455c57533d287e3da20d3110ec6b4fa5d2R89-R90) as part of the profile
(3) When loading the desired profile, [pass home and its error to `expandHomeDir`](https://github.com/snapcore/snapd/pull/13342/files#diff-5862fc64285517730e1ae794ad4f4c455c57533d287e3da20d3110ec6b4fa5d2R121-R123). For non-nil error it would produce an error, _only if where was anything to expand_, that only reference the first mount expansion that fails. This will result in `snap-update-ns` exiting with error.
(4) When assumptions is requested, if there was home error, it [will not unrestrict the home](https://github.com/snapcore/snapd/pull/13342/files#diff-5862fc64285517730e1ae794ad4f4c455c57533d287e3da20d3110ec6b4fa5d2R121-R123).